### PR TITLE
wasm demo completions support

### PR DESCRIPTION
**Pull request by JasperDeSutter**
Adds basic typing completions support to the wasm demo. Things such as functions, the rust-analyzer helpers (box, dbg...) work now.
This includes some refactoring to have the same `Conv` and `ConvWith<CTX>` as used in ra_lsp_server.

### CODE REVIEW

File: wasm-demo/src/conv.rs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                }
+            },
+            documentation: self
+                .documentation()
+                .map(|doc| return_types::MarkdownString { value: doc.as_str().to_string() }),
+            filterText: self.lookup().to_string(),
+            additionalTextEdits: additional_text_edits,
+        }
+    }
+}
```
Comment (matklad) [inline wasm-demo/src/conv.rs]: It feels bad that we need to literally duplicate this between ra_lsp_server and wasm demo, but still, I think its the right thing to do. Wrapping rust-analyzer into some front-end is a ton of boilerplate (b/c the API surface is large), but it is still much easier than implementing the "compiler" itself.

**Comment by matklad** [comment]
@JasperDeSutter  I've also granted you the permissions on this repository, in case I am slow with reviewing things :)

**Comment by JasperDeSutter** [comment]
cool! thanks 😄

File: wasm-demo/src/conv.rs (hunk 1 above)
Comment (JasperDeSutter) [reply wasm-demo/src/conv.rs]: There might be some opportunity for reducing the code duplication, but that would probably require having an entire "lsp-types"-like crate for monaco.
This way rust-analyzer could have an intermediate crate mapping ra-ide to lsp, which both lsp-server and the wasm demo could use.
It sounds like too much effort to me, including the fact that monaco itself isn't stable and changes API regularly and does everything just a little differently.

**Pull request merged**